src/async/pitfalls/pin.md

```diff
@@ -1,13 +1,20 @@
 # `Pin`
 
-When you await a future, all local variables (that would ordinarily be stored on
-a stack frame) are instead stored in the Future for the current async block. If
-your future has pointers to data on the stack, those pointers might get
-invalidated. This is unsafe.
-
-Therefore, you must guarantee that the addresses your future points to don't
-change. That is why we need to "pin" futures. Using the same future repeatedly
-in a `select!` often leads to issues with pinned values.
+All async blocks and async functions return types implementing the `Future`
+trait. When the async code awaits another future, the `Future` trait holds the
+local variables that would normally be located on the stack.
```

Comment:
I don't think this should be linked to await specifically. My understanding is that the async function body is _compiled_ to a struct which `impl Future`. What was local variables in the function body become struct fields.

So what you see as a local variable in the code is never actually a variable in the code after the compiler transformation. Thus, I would not say that this happens when you call `await` since it happened already in the immediate desugaring.

An example of this desugaring can be seen in [Async fn as a Future](https://tokio.rs/tokio/tutorial/async#async-fn-as-a-future). The `State0`, `State1`, and `Terminated` enum variants correspond to different parts of the now-gone async function body. The variants hold the local variables that are alive in the different parts of the body.